### PR TITLE
perf: Eliminate the need for attention DP padding when possible

### PR DESCRIPTION
**Pull request by jinyangyuan-nvidia**
In the current implementation, all ranks are padded to have the same number of tokens after attention when attention DP is used. Therefore, the all-gather, MoE, and reduce-scatter operations potentially process more tokens than necessary, especially when request rate is relatively low.

This PR improves the performance by eliminating the need for attention DP padding when possible. The main idea is to modify `AllgatherOp` and `ReducescatterOp` to support different sizes across different ranks in `gather_dim` and `scatter_dim`, respectively.

Performance Test with DeepSeek-R1-FP8 on 1 node with 8 * H200 (use trtllm-bench throughput: https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/models/core/deepseek_v3#example-multi-node-benchmark-on-gb200-slurm-cluster, the number of tested requests is changed to 15360)
- w/o this PR (the achieved concurrency is 159 requests per GPU)
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     4.9006
Total Output Throughput (tokens/sec):             10036.3364
Total Token Throughput (tokens/sec):              15054.5046
Total Latency (ms):                               3134338.9501
Average request latency (ms):                     578637.2961
Per User Output Throughput [w/ ctx] (tps/user):   3.7876
Per GPU Output Throughput (tps/gpu):              1254.5420
```
- w/ this PR (the achieved concurrency is increased to 168 requests per GPU under the same `kv_cache_free_gpu_mem_fraction`) 
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     5.1881
Total Output Throughput (tokens/sec):             10625.1556
Total Token Throughput (tokens/sec):              15937.7335
Total Latency (ms):                               2960641.8052
Average request latency (ms):                     550466.8627
Per User Output Throughput [w/ ctx] (tps/user):   3.9456
Per GPU Output Throughput (tps/gpu):              1328.1445
```

Performance Test with DeepSeek-R1-FP4 on 1 node with 8 * B200 (use trtllm-bench throughput: https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/models/core/deepseek_v3#example-multi-node-benchmark-on-gb200-slurm-cluster, the number of tested requests is changed to 15360)
- w/o this PR (the achieved concurrency is 384 requests per GPU)
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     17.2031
Total Output Throughput (tokens/sec):             35231.8734
Total Token Throughput (tokens/sec):              52847.8101
Total Latency (ms):                               892864.2431
Average request latency (ms):                     176424.8895
Per User Output Throughput [w/ ctx] (tps/user):   11.7253
Per GPU Output Throughput (tps/gpu):              4403.9842
```
- w/ this PR (the achieved concurrency is also 384) 
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     17.4393
Total Output Throughput (tokens/sec):             35715.6479
Total Token Throughput (tokens/sec):              53573.4719
Total Latency (ms):                               880770.2461
Average request latency (ms):                     174007.1664
Per User Output Throughput [w/ ctx] (tps/user):   11.8895
Per GPU Output Throughput (tps/gpu):              4464.4560
```

### CODE REVIEW

File: cpp/tensorrt_llm/thop/allgatherOp.cpp [hunk 1]
```diff
@@ -56,55 +55,123 @@ class AllgatherOp
...
     {
+        TLLM_CHECK_WITH_INFO(mNcclComm.get() != nullptr, "mNcclComm should be initialized before used");
         auto stream = at::cuda::getCurrentCUDAStream(input.get_device());
+        auto type = tensorrt_llm::runtime::TorchUtils::dataType(input.scalar_type());
         std::vector<int64_t> outputShape = input.sizes().vec();
-        outputShape.insert(outputShape.begin(), mGroup.size());
+        if (all_rank_split_size.has_value())
+        {
+            int64_t total_size = 0;
+            for (auto const& split_size : all_rank_split_size.value())
```
Comment (djns99) [inline cpp/tensorrt_llm/thop/allgatherOp.cpp]: Can we use c++ `std::accumulate()`/`std::reduce()` here?

File: cpp/tensorrt_llm/thop/allgatherOp.cpp [hunk 2]
```diff
@@ -56,55 +55,123 @@ class AllgatherOp
...
+        }
         auto output = torch::empty(outputShape, input.options());
-        size_t size = input.numel();
-        TLLM_CHECK_WITH_INFO(mNcclComm.get() != nullptr, "mNcclComm should be initialized before used");
-        NCCLCHECK(ncclAllGather(
-            input.data_ptr(), output.mutable_data_ptr(), size, (*getDtypeMap())[mType], *mNcclComm, stream));
+        if (all_rank_split_size.has_value())
+        {
+            size_t numel_base = 1;
+            for (auto it = outputShape.cbegin() + 1; it != outputShape.cend(); ++it)
```
Comment (djns99) [inline cpp/tensorrt_llm/thop/allgatherOp.cpp]:
```
size_t numel_base = std::accumulate(outputShape.cbegin() + 1, outputShape.cend(), 1, std::multiplies<>{})
```
(Or equivalently `std::reduce`)

File: cpp/tensorrt_llm/thop/reducescatterOp.cpp [hunk 3]
```diff
@@ -56,55 +55,128 @@ class ReducescatterOp
...
+        }
         auto output = torch::empty(outputShape, input.options());
-        size_t const size = output.numel();
-        TLLM_CHECK_WITH_INFO(mNcclComm.get() != nullptr, "mNcclComm should be initialized before used");
-        NCCLCHECK(ncclReduceScatter(
-            input.data_ptr(), output.mutable_data_ptr(), size, (*getDtypeMap())[mType], ncclSum, *mNcclComm, stream));
+        if (all_rank_split_size.has_value())
+        {
+            size_t numel_base = 1;
+            for (auto it = outputShape.cbegin() + 1; it != outputShape.cend(); ++it)
```
Comment (djns99) [inline cpp/tensorrt_llm/thop/reducescatterOp.cpp]: `std::accumulate`/`std::reduce`

File: cpp/tensorrt_llm/thop/allgatherOp.cpp (hunk 1 above)
Comment (jinyangyuan-nvidia) [reply cpp/tensorrt_llm/thop/allgatherOp.cpp]: Thanks for the valuable suggestion. The code has been modified accordingly.

File: cpp/tensorrt_llm/thop/allgatherOp.cpp (hunk 2 above)
Comment (jinyangyuan-nvidia) [reply cpp/tensorrt_llm/thop/allgatherOp.cpp]: Thanks. Done.

File: cpp/tensorrt_llm/thop/reducescatterOp.cpp (hunk 3 above)
Comment (jinyangyuan-nvidia) [reply cpp/tensorrt_llm/thop/reducescatterOp.cpp]: Thanks. Done.

File: tensorrt_llm/_torch/modules/fused_moe.py [hunk 4]
```diff
@@ -821,10 +797,7 @@ def forward_chunk(
             final_hidden_states = final_hidden_states[0]
 
         if not self.enable_alltoall:
-            if self.reduce_results and self.parallel_size > 1:
-                return self.all_reduce(final_hidden_states)
-            else:
-                return final_hidden_states
+            return final_hidden_states
```
Comment (dongxuy04) [inline tensorrt_llm/_torch/modules/fused_moe.py:800]: Why do we remove `self.all_reduce` here when `self.reduce_results and self.parallel_size > 1`?

Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/modules/fused_moe.py:800]: `self.all_reduce` has been moved to `self.reducescatter_or_allreduce` to better overlap computation and communication when MoE chunking is used.

Comment (dongxuy04) [reply tensorrt_llm/_torch/modules/fused_moe.py:800]: Oh, I see, thanks.

**Review by djns99 — APPROVED** [review]
LGTM

**Review by dongxuy04 — APPROVED** [review]

**Review by juney-nvidia — APPROVED** [review]

**Review by QiJune — APPROVED** [review]

**Review by lucaslie — APPROVED** [review]

**Comment by litaotju** [comment]
@jinyangyuan-nvidia good contributions. Do you have some perf data to show the speedup of this optimization?

That would be really helpful for us to know the perf optimization progress/status and approve this faster if its a big improvement in numbers.

File: cpp/tensorrt_llm/thop/allgatherOp.cpp [hunk 5]
```diff
@@ -56,55 +55,115 @@ class AllgatherOp
...
     std::shared_ptr<ncclComm_t> mNcclComm;
 };
 
 } // namespace
 
 #endif // ENABLE_MULTI_DEVICE
 
-torch::Tensor allgather(torch::Tensor input, torch::List<int64_t> group_)
+torch::Tensor allgather(
+    torch::Tensor input, torch::optional<torch::List<int64_t>> all_rank_split_size, torch::List<int64_t> group_)
```
Comment (litaotju) [inline cpp/tensorrt_llm/thop/allgatherOp.cpp:122]: a nit. How about just name it as "sizes"?

This argument is really long ^-^..

File: cpp/tensorrt_llm/thop/allgatherOp.cpp [hunk 6]
```diff
@@ -56,55 +55,115 @@ class AllgatherOp
...
     return output;
 #else
     return input;
 #endif // ENABLE_MULTI_DEVICE
 }
 
+std::vector<torch::Tensor> allgather_list(torch::TensorList input_list,
+    torch::optional<torch::List<int64_t>> all_rank_split_size, torch::List<int64_t> group_)
+{
+#if ENABLE_MULTI_DEVICE
```
Comment (litaotju) [inline cpp/tensorrt_llm/thop/allgatherOp.cpp:142]: We do not support windows anymore. 

This macro can be safely removed.

File: tensorrt_llm/_torch/auto_deploy/custom_ops/dist.py [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
+
 import torch
 
 from ..distributed import common as dist
 from ..distributed import trtllm as trtllm_dist
 
 
 @torch.library.custom_op("dist::all_gather", mutates_args=(), device_types="cuda")
-def all_gather(tensor: torch.Tensor, dim: int = 0) -> torch.Tensor:
+def all_gather(
```
Comment (litaotju) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/dist.py:12]: The register_fake must be updated to include the new input args.

Otherwise, shape inference (used by torch.compile will fail).

cc @liji-nv @yizhang-nv too. They meet many of this kind of issue.

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 8]
```diff
@@ -457,25 +457,33 @@ def should_enable_alltoall(model_config: ModelConfig, top_k: int) -> bool:
...
-                                          self.mapping,
-                                          gather_dim=0)
+                hidden_states = allgather(
+                    hidden_states,
+                    self.mapping,
+                    gather_dim=0,
+                    all_rank_split_size=all_rank_num_tokens)
+            elif not self.experts.is_cutlass() or (not self.experts.has_fp8_qdq
+                                                   and self.experts.has_nvfp4):
+                # Use padding when not using the cutlass path or when x_sf in self.experts is not None
```
Comment (litaotju) [inline tensorrt_llm/_torch/models/modeling_deepseekv3.py:472]: why cutlass path requires padding?
While the otehr one doesn't? It's tokenwise op right?

File: tests/integration/test_lists/test-db/l0_dgx_h200.yml [hunk 9]
```diff
@@ -18,3 +18,4 @@ l0_dgx_h200:
   # - accuracy/test_llm_api_pytorch.py::TestDeepSeekR1::test_fp8_blockscale[throughput] # OOM
   - accuracy/test_llm_api_pytorch.py::TestDeepSeekR1::test_fp8_blockscale[latency] # 1h
   - unittest/_torch/multi_gpu_modeling/test_llama4.py::test_llama4[pp1-ep1-enable_graph-tp8-trtllm-scout]
+  - unittest/_torch/multi_gpu_modeling -k "deepseek"
```
Comment (litaotju) [inline tests/integration/test_lists/test-db/l0_dgx_h200.yml:21]: This should be covered by h100 test cases already?

File: tests/unittest/_torch/multi_gpu_modeling/test_deepseek.py [hunk 10]
```diff
@@ -23,8 +23,13 @@ def similar(a, b, threshold=0.9):
                          ids=["deepseekv3_lite"])
 @pytest.mark.parametrize("backend", ["TRTLLM"], ids=["trtllm"])
 @pytest.mark.parametrize("quant", ["bf16"])
-@pytest.mark.parametrize("tp_size", [1], ids=["tp1"])
-def test_deepseek_streaming(model_name, backend, quant, tp_size):
+@pytest.mark.parametrize("tp_size", [1, 4], ids=["tp1", "tp4"])
```
Comment (litaotju) [inline tests/unittest/_torch/multi_gpu_modeling/test_deepseek.py:26]: This increased 8 more cases? Can these be covered by some other cases already?

And it does not need to be out product I think

**Review by litaotju — CHANGES_REQUESTED** [review]

File: tests/integration/test_lists/test-db/l0_dgx_h200.yml (hunk 9 above)
Comment (jinyangyuan-nvidia) [reply tests/integration/test_lists/test-db/l0_dgx_h200.yml:21]: I have searched the `tests` path. It seems that deepseek is only tested on 1 GPU before.

```
l0_h100:
- condition:
    ranges:
      system_gpu_count:
        gte: 1
        lte: 1
    wildcards:
      gpu:
      - '*h100*'
      linux_distribution_name: ubuntu*
    terms:
      stage: pre_merge
      backend: pytorch
  tests:
  # ------------- PyTorch tests ---------------
  # Only key models in H100: llama/mixtral/nemotron/deepseek
  - unittest/_torch -k "not (modeling or multi_gpu or auto_deploy)"
  - unittest/_torch -k "modeling_llama"
  - unittest/_torch/multi_gpu_modeling -k "deepseek"
```

File: tests/unittest/_torch/multi_gpu_modeling/test_deepseek.py (hunk 10 above)
Comment (jinyangyuan-nvidia) [reply tests/unittest/_torch/multi_gpu_modeling/test_deepseek.py:26]: I am sure that chunked MoE and the combination of ADP and chunked MoE are not tested in other places. Other cases are added in consideration of completeness. I can remove some cases if you want.

File: tensorrt_llm/_torch/auto_deploy/custom_ops/dist.py (hunk 7 above)
Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/dist.py:12]: It has already been modified in https://github.com/NVIDIA/TensorRT-LLM/pull/3439/files#diff-bd0cf5715e06959a4f9760ca7cf42113b1d76683566a6e827f614c25e6df845fL63-L66

**Comment by jinyangyuan-nvidia** [comment]
> @jinyangyuan-nvidia good contributions. Do you have some perf data to show the speedup of this optimization?
> 
> That would be really helpful for us to know the perf optimization progress/status and approve this faster if its a big improvement in numbers.

The perf data has been added.

File: cpp/tensorrt_llm/thop/allgatherOp.cpp (hunk 5 above)
Comment (jinyangyuan-nvidia) [reply cpp/tensorrt_llm/thop/allgatherOp.cpp]: Done.

File: cpp/tensorrt_llm/thop/allgatherOp.cpp (hunk 6 above)
Comment (jinyangyuan-nvidia) [reply cpp/tensorrt_llm/thop/allgatherOp.cpp]: Done.

File: tests/unittest/_torch/multi_gpu_modeling/test_deepseek.py (hunk 10 above)
Comment (jinyangyuan-nvidia) [reply tests/unittest/_torch/multi_gpu_modeling/test_deepseek.py:26]: Only one extra test case is added now.

File: cpp/tensorrt_llm/thop/allgatherOp.cpp (hunk 6 above)
Comment (jinyangyuan-nvidia) [reply cpp/tensorrt_llm/thop/allgatherOp.cpp]: CI cannot pass after this modification. Will revert the code.

**Review by litaotju — APPROVED** [review]
